Sources/SwiftFusion/Geometry/Pose2.swift

```diff
@@ -35,32 +35,110 @@ import TensorFlow
+
+  public mutating func move(along direction: Coordinate.LocalCoordinate) {
+    coordinateStorage = coordinateStorage.global(direction)
+  }
 
   /// Creates a `Pose2` with rotation `r` and translation `t`.
   ///
   /// This is the bijection SO(2) x R^2 -> SE(2), where "x" means direct product of groups. (Note:
   /// not a group homomorphism!)
   @differentiable
   public init(_ r: Rot2, _ t: Vector2) {
-    tStorage = t
-    rotStorage = r
+    self.init(coordinate: Pose2Coordinate(r, t))
+  }
+  
+  // MARK: Convenience Attributes
+  
+  @differentiable public var t: Vector2 { coordinate.t }
+  
+  @differentiable public var rot: Rot2 { coordinate.rot }
+  
+  /// Twist coordinates, with the first component for rotation and the remaining components for
+  /// translation.
+  ///
+  /// See the documentation comment on the declaration of `Pose2` for more information.
+  public typealias TangentVector = Vector3
+
+  /// Product of two rotations.
+  @differentiable
+  public static func * (lhs: Pose2, rhs: Pose2) -> Pose2 {
+    Pose2(coordinate: lhs.coordinate * rhs.coordinate)
+  }
+  
+  @differentiable
+  public func inverse() -> Pose2 {
+    Pose2(coordinate: coordinate.inverse())
+  }
+}
+
+// MARK: Convenience initializers
+extension Pose2 {
+  /// Creates a `Pose2` with translation `x` and `y` and with rotation `theta`.
+  @differentiable
+  public init(_ x: Double, _ y: Double, _ theta: Double) {
+    self.init(Rot2(theta), Vector2(x, y))
+  }
+
+  public init(randomWithCovariance covariance: Tensor<Double>) {
+    self.init(0, 0, 0)
+    let r = matmul(cholesky(covariance), Tensor<Double>(randomNormal: [3, 1])).scalars
+    let tv = Pose2.TangentVector(r[0], r[1], r[2])
+    self.move(along: tv)
+  }
+}
+
+// MARK: - Global Coordinate System
+
+public struct Pose2Coordinate: Equatable, KeyPathIterable {
+  var tStorage: Vector2
+  var rotStorage: Rot2
+  
+  @differentiable
+  public var t: Vector2 { tStorage }
+  
+  @differentiable
+  public var rot: Rot2 { rotStorage }
+  
+  /// Twist coordinates, with the first component for rotation and the remaining components for
+  /// translation.
+  ///
+  /// See the documentation comment on the declaration of `Pose2` for more information.
+  public typealias TangentVector = Vector3
+}
+
+public extension Pose2Coordinate {
+  @differentiable
+  init(_ rot: Rot2, _ t: Vector2) {
+    self.tStorage = t
+    self.rotStorage = rot
+  }
+
+  /// Returns the rotation (`w`) and translation (`v`) components of `tangentVector`.
+  static func decomposed(tangentVector: Vector3) -> (w: Vector1, v: Vector2) {
+    (Vector1(tangentVector.x), Vector2(tangentVector.y, tangentVector.z))
   }
+  
+  /// Creates a tangent vector given rotation (`w`) and translation (`v`) components.
+  static func tangentVector(w: Vector1, v: Vector2) -> Vector3 {
+    Vector3(w.x, v.x, v.y)
+  }
+}
+
+// MARK: Custom Derivatives
 
+extension Pose2Coordinate {
```

Comment:
Now the interesting stuff starts happening. You should be able to simply delete this entire extension, because the custom derivatives for `init`, `t`, and `rot` are no longer needed.